### ISSUE TITLE
useStyles hook

### ISSUE THREAD

**Issue by matthewp** [matthewp/haunted#121]
This is for discussion, to get feedback if this is something we should do or not. lit-element has a method for setting `adoptedStylesheets`, which is a nice way for doing component CSS.

We could have something similar that works like:

```js
import { css, html } from 'lit-html';
import { component, useStyles } from 'haunted';

function MyApp() {
  useStyles(() => css`header { background: green; }`);

  return html`
    <header><h1>Hello world</h1></header>
  `;
}
```

Something like this. This would probably need [useLayoutEffect](https://github.com/matthewp/haunted/issues/120) to have the right timing.

Is this something we should do? Perhaps after examining bundle size effect a bit?

**Comment by threepointone** [comment]
I don't think you should do this as a hook https://overreacted.io/why-isnt-x-a-hook/ In general, things that 'mutate' the tree and change the context below them should probably be components. It's also why we don't have a `useProvider` hook. I would recommend a component like `<Styles>` instead.

**Comment by matthewp** [comment]
Hey thanks! That's great insight. Just for clarity, this would not mutate the tree in supporting browsers (but would in others), but rather set `adoptedStylesheets`. It would be exactly equivalent to:

```js
useLayoutEffect(() => {
  this.adoptedStylesheets = [ css`header { }` ];
});
```

Probably it's a better idea to first do #120 and play around with the idea manually to make sure it feels right.

**Comment by threepointone** [comment]
Setting `adoptedStylesheets` would count as mutation to me, changing the 'behaviour' of children. Styles are like context in that way, at least to me.

**Comment by alfredosalzillo** [comment]
Why not think to something like react styled components ? I was working on this https://github.com/alfredosalzillo/masquerad

**Comment by matthewp** [comment]
Is that repo private @alfredosalzillo? I can't see it.

**Comment by alfredosalzillo** [comment]
@matthewp wrong link, I fix it

**Comment by matthewp** [comment]
@alfredosalzillo Thanks, looks cool! As far as haunted I don't think I want to implement a full CSS solution like that. I was thinking something that only sets adoptedStyleSheets. But talking it through here and the point made by @threepointone I think it's premature to do even a useStyles hook. I'm going to start by implementing useLayoutEffect and see what happens from there.

**Comment by Gladear** [comment]
I would really enjoy having Constructable Stylesheets integrated in **haunted**.
Maybe instead of being a hook, it can be an option in the `component` function ?
Anyway, _lit-html_ does not currently export the `css` template litteral, it's specific to _lit-element_. This seems to be in discussion [here](https://github.com/Polymer/lit-html/issues/823)

**Comment by matthewp** [comment]
Thanks @Gladear, I noticed that as well. For phase 1 I'm going to implement #120 which will provide the necessary underpinnings. I'll post a snippet/gist of what `useStyles` would look like using that. If lit-html adds `css` and people find it useful I could see `useStyles` coming into core in the future.

**Comment by justinfagnani** [comment]
I'm unsure about moving LitElement's `css` tag to lit-html, vs trying to make an even more general adoptedStyleSheets polyfill so everyone can interoperate without a common dependency. `css` might still be useful as a declarative dev interface to that. Keep in touch on that point.

As for @threepointone's suggestion, in lit-html we don't have components - it's just HTML and custom elements are the components if there are any. So there's no equivalent to a `<Style>` component that can be read out of VDOM to apply to adoptedStyleSheets. I don't see why a hook isn't appropriate for this case personally.

**Comment by bvkimball** [comment]
I would really want this as well.  I have been trying to hack around with it.

my first attempt, didn't go so well but:
https://github.com/bvkimball/haunted-styles

I started using [stylis](https://github.com/thysultan/stylis.js) instead of basic css which i think is better for the end-user you can see that here: https://github.com/bvkimball/niht-stylis

I have a fork of haunted were I changed the [render](https://github.com/matthewp/haunted/blob/master/lib/container.js#L75) function to process an array of TemplateResults or CSSResults so my components could look like this:

``` javascript
import { define, html, css, useState } from "haunted";

  function Counter() {
    return ({ test }) => {
      const [count, setCount] = useState(0);
      return [
        css`
          display: inline-flex;
          flex-flow: column;
          text-align: center;
          background: ${count > 10 ? "green" : "red"};
        `,
        html`
          <div id="count">${count}</div>
          <button type="button" @click=${() => setCount(count + 1)}">Increment</button>
        `
      ];
    };
  }

  define("my-counter", Counter);
```

this was just my person preference but thought i would add to the discussion.

**Comment by alfredosalzillo** [comment]
I think an useStyle hook should be implemented like this:

```javascript
import * as stylis from 'stylis';

const uniqueClassName = (css) => {
  // generate the unique class name using for example the murmur hash algorithm
}
const adoptSheet = (el, sheet) => {
  const root = el.getRootNode();
  if (root) {
    root.adoptedStyleSheets = [
      ...new Set([
        ...root.adoptedStyleSheets,
        sheet,
      ]),
    ];
  }
};

const removeSheet =(el, sheet) => {
 const root = el.getRootNode();
  if (root) {
    root.adoptedStyleSheets =root.adoptedStyleSheet.filter(s => s !== sheet);
  }
}

// An hook to retrieve the 'this' of the component
const useHost = hook(class HostHook extends Hook {
  constructor(id, el) {
    super(id, el);
    this.el = el;
  }

  update() { return this.el.host; }
});

const useStyle = (css, values) => {
    const el = useHost();
    const sheet = useMemo(() => new CSSStyleSheet());
    // add the sheet to the adopted stylesheet of the root of the hook's host, and remove on teardown
    useEffect(() => {
       adoptSheet(el, sheet);
       return () => removeSheet(el, sheet);
    }, [el, sheet]);
   const className = uniqueClassName(css);
   useEffect(() => {
      // add the rule to the sheet
      sheet.insertRule(`@media {${stylis(`.${className}`, css)}}`);
      // add the className to the host, and remove on teardown
      el.classList.add(className);
      return () => el.classList.remove(className);
   }, [className, el, sheet]);
}
```
And used this way:

```javascript

import { define, html, css, useState } from "haunted";

  function Counter() {
    return ({ test }) => {
      const [count, setCount] = useState(0);
      useStyle(`
          display: inline-flex;
          flex-flow: column;
          text-align: center;
          background: ${count > 10 ? "green" : "red"};
        `, [count]);
      return html`
          <div id="count">${count}</div>
          <button type="button" @click=${() => setCount(count + 1)}">Increment</button>
        `;
    };
  }

  define("my-counter", Counter);
```

**Comment by matthewp** [comment]
@alfredosalzillo Why a unique style name? In shadow DOM that shouldn't be required I don't think.

**Comment by matthewp** [comment]
@bvkimball We need #120 to make this happen, then it should be fairly trivial to do this yourself. I'll work on that issue some time this week.

**Comment by cletusw** [comment]
Now that `useLayoutHook` is merged (#136) how might an implementation of this look? Thanks!

**Comment by andyinabox** [comment]
There's a hook implementation here that seems worth a look: https://whoisryosuke.com/blog/2020/adding-constructable-stylesheets-to-hauntedjs/
